monai/metrics/cumulative_average.py

```diff
@@ -9,64 +9,168 @@
 
-import torch
-
-from monai.transforms import isnan
-from monai.utils import convert_data_type
+import warnings
+from typing import Any, Optional, Union
 
-from .metric import Cumulative
+import torch
+import torch.distributed as dist
+from numpy.typing import NDArray
+from torch.distributed import ProcessGroup
 
 
-class CumulativeAverage(Cumulative):
+class CumulativeAverage:
     """
-    Cumulatively record data value and aggregate for the average value.
-    It supports single class or multi-class data, for example,
-    value can be 0.44 (a loss value) or [0.3, 0.4] (metrics of two classes).
-    It also supports distributed data parallel, sync data when aggregating.
-    For example, recording loss values and compute the overall average value in every 5 iterations:
+    A utility class to keep track of average values. For example during training/validation loop,
+    we need to accumulate the per-batch metrics and calculate the final average value for the whole dataset.
+    When training in multi-gpu environment, with DistributedDataParallel, it will average across the processes.
+
+    Example:
 
     .. code-block:: python
 
-        average = CumulativeAverage()
-        for i, d in enumerate(dataloader):
-            loss = ...
-            average.append(loss)
-            if i % 5 == 0:
-                print(f"cumulative average of loss: {average.aggregate()}")
-        average.reset()
+        from monai.metrics import CumulativeAverage
+
+        run_avg = CumulativeAverage()
+        for i in range(len(train_set)):
+            ...
+            val = calc_metric(x,y) #some metric value
+            run_avg.append(val)
+
+        val_avg = run_avg.aggregate() #average value
 
     """
 
-    def __init__(self) -> None:
-        super().__init__()
-        self.sum = None
-        self.not_nans = None
+    def __init__(self, group: Optional[ProcessGroup] = None) -> None:
+        """
+        Args:
+            group: Optional process group to use in multi-gpu distributed setup. Defaults to None
+            to use the default process group created during DistributedDataParallel initialization.
+        """
 
-    def reset(self):
+        self.is_distributed = dist.is_available() and dist.is_initialized()
+        self.device = None
+
+        if self.is_distributed:
+            self.group = group
+            if dist.get_backend(group=group) == dist.Backend.NCCL:
+                self.device = torch.device("cuda", dist.get_rank(group=group))
+        else:
+            self.group = None
+
+        if self.device is None:
+            self.device = torch.device("cpu")
+
+        self.reset()
+
+    def reset(self) -> None:
         """
-        Reset all the running status, including buffers, sum, not nans count, etc.
+        Reset all the running status
 
         """
-        super().reset()
-        self.sum = None
-        self.not_nans = None
+        self.val: torch.Tensor = 0  # type: ignore
```

Comment:
It's slightly inefficient but using actual tensors here instead of 0 would be more consistent and not need the type ignore comment, but it's not critical.

Reply:
yes, I tried it, using actual torch tensor 0, but we don't know the shape of val, sum, count before stating to accumulate, and later self.sum += some_tensor fails, since in place increment not possible with different tensor shapes (even if incrementing 0 dim tensor).  So we'd have to use self.sum = self.sum + some_tensor, out of place increment, which is slightly bigger overhead. It's seems simpler to have it as is.